contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import {CoreRef} from "../refs/CoreRef.sol";
+import {TribeRoles} from "./../core/TribeRoles.sol";
+import {RateLimited} from "./RateLimited.sol";
+import {IMultiRateLimited} from "./IMultiRateLimited.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+/// @title contract for putting a rate limit on how fast an address can perform an action e.g. Minting
+/// there are two buffers, one buffer which is each individual addresses's current buffer,
+/// and then there is a global buffer which is the buffer that each individual address must respect as well
+/// @author Fei Protocol
+contract MultiRateLimited is RateLimited, IMultiRateLimited {
+    using SafeCast for *;
+
+    /// @notice the struct containing all information per rate limited address
+    struct RateLimitData {
+        uint32 lastBufferUsedTime;
+        uint144 bufferCap;
+        uint112 rateLimitPerSecond;
+        uint144 bufferStored;
+    }
+
+    /// @notice rate limited address information
+    mapping (address => RateLimitData) public rateLimitPerAddress;
+
+    /// @notice max rate limit per second allowable by non governor
+    uint256 public maxRateLimitPerSecond;
+
+    /// @notice max buffer cap allowable by non governor
+    uint256 public maxBufferCap;
+
+    constructor(
+        address coreAddress,
+        uint256 _maxRateLimitPerSecond,
+        uint256 _rateLimitPerSecond,
+        uint256 _maxRateLimitPerSecondMRL,
+        uint256 _maxBufferCap,
+        uint256 _globalBufferCap,
+        bool _doPartialAction
+    )
+        CoreRef(coreAddress)
+        RateLimited(_maxRateLimitPerSecond, _rateLimitPerSecond, _globalBufferCap, _doPartialAction)
+    {
+        require(_maxBufferCap < _globalBufferCap, "MultiRateLimited: max buffer cap invalid");
+
+        maxRateLimitPerSecond = _maxRateLimitPerSecondMRL;
+        maxBufferCap = _maxBufferCap;
+    }
+
+    // ----------- Governor and Admin only state changing api -----------
+
+    /// @notice update the ADD_MINTER_ROLE rate limit per second
```

Comment:
Is this supposed to be a directive "you need to update the ADD_MINTER_ROLE..." or did you just miss an s, ie it should be "updates"?

Reply:
(closing, will fix in my nitpicks pr)